packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts

```diff
@@ -0,0 +1,301 @@
+     */
+    lockTimeoutAfterSuccess?: number
+  }
+  /**
+   * If true, the job will log when it starts and finishes.
+   */
+  shouldLogExecution?: boolean
+}
+
+export type LockConfiguration = {
+  lockSuffix?: string
+  identifier?: string
+  refreshInterval?: number
+  lockTimeout?: number
+  acquiredExternally?: true
+}
+
+export function createTask(logger: CommonLogger, job: AbstractPeriodicJob) {
+  const executorId = randomUUID()
+
+  logger.info({
+    msg: 'Periodic job registered',
+    jobId: job.options.jobId,
+    executorId,
+  })
+
+  return new AsyncTask(
+    job.options.jobId,
+    () => {
+      return job.process(executorId)
+    },
+    /* v8 ignore next 9 */
+    (error) => {
+      logger.error(
+        stdSerializers.err({
+          name: error.name,
+          message: error.message,
+          stack: error.stack,
+        }),
+      )
+    },
+  )
+}
+
+export type PeriodicJobDependencies = {
+  redis: Redis
+  logger: CommonLogger
+  transactionObservabilityManager: TransactionObservabilityManager
+  errorReporter: ErrorReporter
+  scheduler: ToadScheduler
+}
+
+export abstract class AbstractPeriodicJob {
+  public readonly options: Required<BackgroundJobConfiguration>
+  protected readonly redis: Redis
+  protected readonly transactionObservabilityManager: TransactionObservabilityManager
+  protected readonly logger: CommonLogger
+  protected readonly errorReporter: ErrorReporter
+  protected readonly scheduler: ToadScheduler
+  private singleConsumerLock?: Mutex
+
+  protected constructor(
+    options: BackgroundJobConfiguration,
+    {
+      redis,
+      logger,
+      transactionObservabilityManager,
+      errorReporter,
+      scheduler,
+    }: PeriodicJobDependencies,
+  ) {
+    this.options = {
+      intervalInMs: DEFAULT_JOB_INTERVAL,
+      shouldLogExecution: false,
+      ...options,
+      singleConsumerMode: {
+        enabled: false,
+        exclusiveLockSuffix: options.singleConsumerMode?.exclusiveLockSuffix ?? 'EXCLUSIVE',
+        lockTimeout: (options.intervalInMs ?? DEFAULT_JOB_INTERVAL) * 2,
+        ...options.singleConsumerMode,
+      },
+    }
+    this.logger = logger
+    this.redis = redis
+    this.transactionObservabilityManager = transactionObservabilityManager
+    this.errorReporter = errorReporter
+    this.scheduler = scheduler
+  }
+
+  public register() {
+    const task = createTask(this.logger, this)
+
+    this.scheduler.addSimpleIntervalJob(
+      new SimpleIntervalJob(
+        {
+          milliseconds: this.options.intervalInMs,
+          runImmediately: true,
+        },
+        task,
+        {
+          id: this.options.jobId,
+          preventOverrun: true,
+        },
+      ),
+    )
+  }
+
+  public async dispose() {
+    this.scheduler.stopById(this.options.jobId)
+    await this.singleConsumerLock?.release()
+  }
+
+  protected abstract processInternal(executionUuid: string): Promise<unknown>
+
+  public async process(executorId: string) {
```

Comment:
On `AbstractBackgroundJobProcessor` we use these names but in the other direction, should we follow the same approach? (`process` for the abstract method and `processInternal` for the private one)

Reply:
not sure. note that there needs to be a public entry point that toad-scheduler task can invoke. would you call that `processInternal`?

Reply:
Oh I see, sorry for the misunderstanding, wondering if we can make the method private and `createTask`  receive the method to call, something like:
```
createTask(logger: CommonLogger, jobId: string, method: (execId: string) => Promise<void>): AsyncTask
```

But I see createTask is exposed, so maybe it is not a good idea 🤔

Reply:
yeah, I would probably keep it as-is, not sure that tinkering with this part would decrease complexity significantly

Reply:
Yeah, I am fine with it, my only point was not allowing to directly call that method or override it